Data/IntMap/Internal.hs

```diff
@@ -937,6 +944,48 @@ updateLookupWithKey _ _ Nil = (Nothing,Nil)
 -- | /O(min(n,W))/. The expression (@'alter' f k map@) alters the value @x@ at @k@, or absence thereof.
 -- 'alter' can be used to insert, delete, or update a value in an 'IntMap'.
 -- In short : @'lookup' k ('alter' f k m) = f ('lookup' k m)@.
+
+#if __GLASGOW_HASKELL__ >= 802
+alter :: (Maybe a -> Maybe a) -> Key -> IntMap a -> IntMap a
+alter f !k t = case alter# f k t of
+  (# (# #) | #) -> t
+  (# | t' #) -> t'
+{-# INLINE alter #-}
+
+
+-- Internal implementation which keeps track of whether or not the intmap was
+-- modified using an unboxed sum (Maybe).
+--
+-- If no modifications are made to the map (# (# #) | #) is returned, otherwise
+-- (# | newMap #) is returned.
+alter# :: (Maybe a -> Maybe a) -> Key -> IntMap a -> (# (# #) | IntMap a #)
```

Comment:
What happens if you make the function have type `Maybe# a -> Maybe# a`, where `Maybe#` is the unboxed version of `Maybe`? Couldn't we often avoid the `Maybe` allocation that way? Benchmark!

Reply:
That function is user defined (passed to `alter`) so at some point you would have to do the conversion from `Maybe` to `Maybe#`. As far as I know coerce wouldn't work for `f` because they have different representations (`Maybe` and `Maybe#` that is), but there may be another way that I'm unaware of. Did you have a specific approach in mind?

Reply:
I think it's a decent bet that the function we're passed will be small enough to inline. So calling `alter#` with

```haskell
\ p -> fromMaybe (f (toMaybe p))
```

(essentially) should usually avoid any actual `Maybe`s. Or so I imagine.